### ISSUE TITLE
Timeout occurs during build

### ISSUE THREAD

**Issue by hellow554** [integer32llc/rust-playground#712]
It seems to me, that the backend is not fast enough to actually compile the code anymore.

It seems to be random, therefore maybe load dependent, but I sometimes get a timeout even with a simple `fn main() {}`.

Can you investigate that? I'm not sure how to reproduce it, because as said, it is somewhat random and does not depend on the code itself.

**Comment by shepmaster** [comment]
Yep. I think it's been this way for a few months now. I haven't had time to figure it out, or even to add some basic metrics to help in figuring it out. It _may_ just be that Rust / Cargo / the various crates are slower.

<img width="1647" alt="image" src="https://user-images.githubusercontent.com/174509/116631927-f1c37480-a923-11eb-8539-c5fab4e2535b.png">

<img width="1663" alt="image" src="https://user-images.githubusercontent.com/174509/116631962-0273ea80-a924-11eb-9d70-b066ec5b11ad.png">

It's a 2x Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz and ~8GiB of RAM.

**Comment by josephlr** [comment]
I've also been hitting this recently, but running with `miri` usually doesn't hit this. Is running w/ Miri less CPU intensive?

**Comment by Bauxitedev** [comment]
I'm getting timeouts repeatedly now. I can hardly run any code in the playground, even just basic hello worlds don't run anymore.

**Comment by shepmaster** [comment]
(responding to https://github.com/integer32llc/rust-playground/issues/762#issue-1090649922)

> To fix this, the timeout duration could be increased from 10 seconds to a higher value.

I'm not well-versed in queuing theory, but wouldn't this make it worse or otherwise not help? If we assume a consistent rate of requests and we increase the timeout, then more compilations will be occurring concurrently, fighting each other for the same resources, slowing each other down for a longer period.   

I also expect that at some inflection point, people will assume "oh, something broke" while waiting for a request and will resubmit the request, exacerbating the problem as now that one user has doubled their load. Even preventing a second submit in the UI is problematic as people can reload the page and retry, which is even a tiny bit more expensive.

> Alternatively, `cargo build` and `cargo run` could have different timeouts, for example 15 seconds for compilation and 2 seconds for running.

This is perhaps something that we should have otherwise, but I also don't see how it would fix the problem. Compilation and running both contend for the same CPU time slice.

**Comment by shepmaster** [comment]
For a picture, here is "Timeouts vs Load average":

<img width="1655" alt="image" src="https://user-images.githubusercontent.com/174509/147698282-2fcb8677-7824-42e6-9ef9-2726fc7b2320.png">

**Comment by shepmaster** [comment]
And for the last 7 days

<img width="1645" alt="timesouts and load average" src="https://user-images.githubusercontent.com/174509/166996019-d5d03080-9180-4f64-a6ad-bd33d3e67acb.png">
